### PR TITLE
Only write required columns to changesets table in bulkProcessor

### PR DESCRIPTION
**Pull request by mrnugget**
This addresses some of the problems described in https://github.com/sourcegraph/sourcegraph/issues/22915, especially in this comment: https://github.com/sourcegraph/sourcegraph/issues/22915#issuecomment-889241304

It doesn't fix the problems completely, but it improves the situation by lessening the change of overwrites that lead to a loss of critical data.

What this PR specifically does:
* Introduce two new methods to the `*Store` that allow updating only single columns: `ui_publication_state` or `batch_change_ids`
* In the `(*bulkProcessor)` it uses them in `detach` and `publishChangeset` to make sure that no other state of the changeset is overwritten
* It adds a sanity check to the `(*bulkProcessor).process` method to ensure  that no changeset that's currently being processes by the reconciler is touched. There should be no good usecase where allowing that to go through makes sense, so I'd rather err on the side of safety here.

In other words: the changes here are a first step towards fixing the problems described in #22915.

### CODE REVIEW

**Review by eseliger — APPROVED** [review]
>  There should be no good usecase where allowing that to go through makes sense, so I'd rather err on the side of safety here.

The comment action was actually ok to do when it is processing.

But not a big change, I think.

File: enterprise/internal/batches/background/bulk_processor.go [hunk 1]
```diff
@@ -51,6 +53,11 @@ func (b *bulkProcessor) process(ctx context.Context, job *btypes.ChangesetJob) (
 		return errors.Wrap(err, "loading changeset")
 	}
 
+	// Changesets that are currently processing should be retried at a later stage.
+	if b.ch.ReconcilerState == btypes.ReconcilerStateProcessing {
+		return changesetIsProcessingErr
+	}
```
[the remark this answers is not in the document]
Comment (mrnugget) [reply enterprise/internal/batches/background/bulk_processor.go:59]: > While something is processing it'd be picked again and again, should we raise the cooldown time between runs a bit? It's currently 5s.

I think that's fine. If something is processing for >10s I don't think it's a problem that we try twice.

> There is no guarantee that this constraint still holds when we are done with the code host action, is this just meant to be an early bail-out?

Exactly. It's not meant to be a lock. Only as a precaution for the cases where we know that it doesn't make sense to do the processing, because we know that something is currently changing.

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/batches/background/bulk_processor.go:59]: > I think that's fine. If something is processing for >10s I don't think it's a problem that we try twice.

But the max retires would be exhausted more quickly than they need to, maybe.

**Comment by mrnugget** [comment]
> The comment action was actually ok to do when it is processing.

Yeah, I know. I just think it's easier to have a rule like this apply for all operations vs. special rules for each operation.

File: enterprise/internal/batches/background/bulk_processor.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrnugget) [reply enterprise/internal/batches/background/bulk_processor.go:59]: We have 10 retries. So if someone applies a new batch spec while someone else creates a bulk processing operation and the bulk processor picks up a changeset that is already processing and processing that changesets takes longer than 50s then it fails. And the user can still create another bulk operation. I'd rather have this happen than something running into the background for minutes without anything showing up.

[the remark this answers is not in the document]
Comment (eseliger) [reply enterprise/internal/batches/background/bulk_processor.go:59]: 👍
